blog: Add "Multiplayer in Godot 4.0: Scene Replication" (#562)
godotengine/godot-website: collections/_article/multiplayer-in-godot-4-0-scene-replication.md

---
title: "Multiplayer in Godot 4.0: Scene Replication"
excerpt: "Create multiplayer games in an instance (pun intended) with the new MultiplayerSpawner and MultiplayerSynchronizer nodes.
Check out the key concepts, and get started with a quick tutorial on how to make a simple game using Godot multiplayer features!"
categories: ["progress-report"]
author: Fabio Alessandrelli
image: /storage/blog/covers/multiplayer-in-godot-4-0-scene-replication.png
date: 2023-02-18 18:00:00
---

> Howdy Godotters! It's been more then a year since I last wrote on this blog, "... blah blah, blah blah, blah" (cf. C. L. C. Chuckie), and with Godot 4.0 nearing release, it s time to write something to jumpstart you into creating your own multiplayer experience with Godot.

Reviewer: is `"... blah blah, blah blah, blah" (cf. C. L. C. Chuckie)` a reference to something? or is it a placeholder that you need to fill out 😅

Author: A sneaky reference https://en.wikipedia.org/wiki/LeChuck , I can remove it if you feel it's too cryptic ;)

Reviewer: haha no, that's fine! I just wasn't sure 😅 
It's been too long since I played Monkey Island to even catch the reference

> Howdy Godotters! It's been more then a year since I last wrote on this blog, "... blah blah, blah blah, blah" (cf. C. L. C. Chuckie), and with Godot 4.0 nearing release, it s time to write something to jumpstart you into creating your own multiplayer experience with Godot.

Reviewer: ```suggestion
Howdy Godotters! It's been more than a year since I last wrote on this blog, "... blah blah, blah blah, blah" (cf. C. L. C. Chuckie), and with Godot 4.0 nearing release, it s time to write something to jumpstart you into creating your own multiplayer experience with Godot.
```

> Since the [last blog post](https://godotengine.org/article/multiplayer-changes-godot-4-0-report-4), we rewrote the scene replication API to be more user-friendly, addeding two "configuration" nodes:

Reviewer: ```suggestion
Since the [last blog post](https://godotengine.org/article/multiplayer-changes-godot-4-0-report-4), we rewrote the scene replication API to be more user-friendly, adding two "configuration" nodes:
```

	start_game()

func start_game():
	# Hide the UI and unpause to start the game.
	$UI.hide()
	get_tree().paused = false
```

Then setup our world scene with some physics objects, and add it to the multiplayer scene:

![Scene Tree and 3D view of the multiplayer scene](/storage/blog/multiplayer-in-godot-4-0-scene-replication/multiplayer-level.png)

Synchronizing Properties
---

So, now that we have our scene set up and peers can connect, let's add a MultiplayerSynchronizer to our object scene and configure it to sync some of its properties.

Synchronized properties can be configured by selecting them from a list or entering their path.

![Editor view of a scene with a MultiplayerSynchronizer](/storage/blog/multiplayer-in-godot-4-0-scene-replication/multiplayer-synchronizer.png)

Running the game now you will notice that the objects start synchronizing as soon as the client connects.

*Tip*: You can run and debug multiple instances simultaneously from the editor by changing the value in `Debug -> Run Multiple Instances`

![Editor debug menu](/storage/blog/multiplayer-in-godot-4-0-scene-replication/debug-multiple-instances.png)

Spawning and despawning scenes
---

> The MultiplayerSpawner node automate the process of replicating dynamically instantiated nodes across peers including when joining mid-game or reconnecting.

Reviewer: ```suggestion
The MultiplayerSpawner node automates the process of replicating dynamically instantiated nodes across peers including when joining mid-game or reconnecting.
```

This mostly boils down to the following MultiplayerSpawner properties:

- `spawn_path`: Defines the path where the spawner will add the nodes it instantiates.

> - `Auto Spawn List`: Defines the scenes to be automatically replicated when added as child of `spawn_path` by the authority (server by default).

Reviewer: ```suggestion
- `Auto Spawn List`: Defines the scenes to be automatically replicated when added as a child of `spawn_path` by the authority (server by default).
```

```
func start_game():
	# Hide the UI and unpause to start the game.
	$UI.hide()
	get_tree().paused = false
	# Only change level on the server.
	# Clients will instantiate the level via the spawner.
	if multiplayer.is_server():
		change_level.call_deferred(load("res://level.tscn"))

# Call this function deferred and only on the main authority (server).
func change_level(scene: PackedScene):
	# Remove old level if any.
	var level = $Level
	for c in level.get_children():
		level.remove_child(c)
		c.queue_free()
	# Add new level.
	level.add_child(scene.instantiate())

# The server can restart the level by pressing HOME.
func _input(event):
	if not multiplayer.is_server():
		return
	if event.is_action("ui_home") and Input.is_action_just_pressed("ui_home"):
		change_level.call_deferred(load("res://level.tscn"))
```

Now the level will be instantiated by the server as soon as it starts, and clients will instantiate it as soon as they connect.

The server can also call `change_level` at any moment to restart the current level, or to select a different one (as long as the selected level is in the "Auto Spawn List" of all peers).

> Additionally, when scenes controlled by a multiplayer spawner contain a multiplayer synchronizer referencing the root node of the scene, the configured "Spawn" properties will be automatically set on remote peers during the spawn process.

Reviewer: ```suggestion
Additionally, when scenes controlled by a multiplayer spawner contain a multiplayer synchronizer referencing the root node of the scene, the configured "Spawn" properties will be automatically set on remote peers during the spawning process.
```

```
# level.gd
extends Node3D

const SPAWN_RANDOM := 5.0

func _ready():
	# We only need to spawn players on the server.
	if not multiplayer.is_server():
		return

	multiplayer.peer_connected.connect(add_player)
	multiplayer.peer_disconnected.connect(del_player)

	# Spawn already connected players
	for id in multiplayer.get_peers():
		add_player(id)

	# Spawn the local player unless this is a dedicated server export.
	if not OS.has_feature("dedicated_server"):
		add_player(1)

func _exit_tree():
	if not multiplayer.is_server():
		return
	multiplayer.peer_connected.disconnect(add_player)
	multiplayer.peer_disconnected.disconnect(del_player)

func add_player(id: int):
	var character = preload("res://player.tscn").instantiate()
	# Set player id.
	character.player = id
	# Randomize character position.
	var pos := Vector2.from_angle(randf() * 2 * PI)
	character.position = Vector3(pos.x * SPAWN_RANDOM * randf(), 0, pos.y * SPAWN_RANDOM * randf())
	character.name = str(id)
	$Players.add_child(character, true)

func del_player(id: int):
	if not $Players.has_node(str(id)):
		return
	$Players.get_node(str(id)).queue_free()
```

> We can now run the game and test that connecting to a host will correctly spawn the appopriate level and players, giving us control to our character, and properly synchronizing the object positions.

Reviewer: ```suggestion
We can now run the game and test that connecting to a host will correctly spawn the appropriate level and players, giving us control of our character, and properly synchronizing the object positions.
```
